File: src/effectscan.c
/**********************************************************************
 * 
 * effectscan.c
 *
 * copyright (c) 2007-8, Karl W Broman 
 *
 * last modified Jan, 2008
 * first written Sep, 2007
 *
 *     This program is free software; you can redistribute it and/or
 *     modify it under the terms of the GNU General Public License,
 *     version 3, as published by the Free Software Foundation.
 * 
 *     This program is distributed in the hope that it will be useful,
 *     but without any warranty; without even the implied warranty of
 *     merchantability or fitness for a particular purpose.  See the GNU
 *     General Public License, version 3, for more details.
 * 
 *     A copy of the GNU General Public License, version 3, is available
 *     at http://www.r-project.org/Licenses/GPL-3
 *
 * C functions for the R/qtl package
 *
 * These functions are for calculating the estimated effects, by multiple
 * imputation, in a single-QTL scan along a chromosome.
 *
 * Contains: R_effectscan, effectscan
 *
 **********************************************************************/

#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <R.h>
#include <Rmath.h>
#include <R_ext/PrtUtil.h>
#include <R_ext/Applic.h>
#include <R_ext/Lapack.h>
#include "effectscan.h"
#include "util.h"

/* R_effectscan: wrapper for effectscan */

void R_effectscan(int *nind, int *ngen, int *ndraws, int *npos,
		  int *draws, double *pheno, double *effectmapping,
		  double *beta, double *se, int *getse)
{
  int ***Draws;
  double **Beta, **SE;

  reorg_errlod(*ngen, *npos, beta, &Beta);
  reorg_errlod(*ngen, *npos, se, &SE);
  reorg_draws(*nind, *npos, *ndraws, draws, &Draws);

  effectscan(*nind, *ngen, *ndraws, *npos, Draws, pheno, effectmapping,
	     Beta, SE, *getse);
}
  

/**********************************************************************
 * effectscan
 *
 * nind   Number of individuals
 * ngen   Number of genotypes
 * ndraws Number of imputations
 * npos   Number of positions
 * Draws  The imputed genotypes (dim nind x npos x ndraws)
 * pheno  Phenotypes (length nind)
 * effectmapping  Matrix of size ngen x ngen, giving the design matrix 
 *                for each possible genotype
 * Beta   On exit, the estimated coefficients (dim npos x ngen)
 * SE     On exit, the estimated standard errors (dim npos x ngen)
 * getse  If 1, calculate SEs; if 0, don't
 *
 **********************************************************************/

void effectscan(int nind, int ngen, int ndraws, int npos,
		int ***Draws, double *pheno, double *mapping,
		double **Beta, double **SE, int getse)
{
  int nphe=1, lwork, info, i, j, s, k, ntrim, *index, *ng, *flag;
  double *resid, *dwork, *var, sigmasq=0.0, *x;
  double *wbeta, *wvar, *wrss, *wts, totwt=0.0;
  double lrss0, mpheno;
  
  lwork = 4*nind;
  ntrim = (int) floor( 0.5*log(ndraws)/log(2.0) );

  allocate_double(nind, &resid);
  allocate_double(lwork, &dwork);
  allocate_double(ngen*ndraws, &var);
  allocate_double(ngen*ndraws, &wbeta);
  allocate_double(ngen*ndraws, &wvar);
  allocate_double(ndraws, &wrss);
  allocate_double(ndraws, &wts);
  allocate_double(ngen*nind, &x);
  allocate_int(ndraws, &index);
  allocate_int(ngen, &ng);
  allocate_int(ndraws, &flag);
  
  /* log null RSS */
  lrss0 = mpheno = 0.0;
  for(i=0; i<nind; i++) mpheno += pheno[i];
  mpheno /= (double)nind;
  for(i=0; i<nind; i++) lrss0 += (pheno[i] - mpheno)*(pheno[i] - mpheno);
  lrss0 = log(lrss0);

  for(s=0; s<npos; s++) { /* loop over positions */

    for(i=0; i<ndraws; i++) { /* loop over imputations */

      /* form X matrix */
      for(k=0; k<ngen; k++) 
	for(j=0; j<nind; j++) 
	  x[j+k*nind] = mapping[Draws[i][s][j]+k*ngen];
	  
      memcpy(resid, pheno, nind*sizeof(double));


      /* look for 0's in the genotype matrix */
      flag[i] = 0;
      for(j=0; j<ngen; j++) ng[j] = 0;
      for(j=0; j<nind; j++) ng[Draws[i][s][j]]++;
      for(j=0; j<ngen; j++) if(ng[j]==0) flag[i] = 1;

      if(!flag[i]) {
	/* linear regression */
	F77_CALL(dgels)("N", &nind, &ngen, &nphe, x, &nind, resid, &nind,
			dwork, &lwork, &info);

        /* coefficient estimates */
        for(j=0; j<ngen; j++) wbeta[j+i*ngen] = resid[j];

        /* residual sum of squares and residual variance */
        wrss[i] = 0.0;
        for(j=ngen; j<nind; j++) wrss[i] += (resid[j]*resid[j]);
        if(getse) sigmasq = wrss[i] / (double)(nind-ngen);

        wts[i] = ((double)nind / 2.0) * (lrss0 - log(wrss[i]));
      }

      if(i==0) totwt = wts[i];
      else totwt = addlog(totwt, wts[i]);
    
      if(getse && !flag[i]) {
	for(j=0; j<ngen; j++)
	  memcpy(var+j*ngen, x+j*nind, ngen*sizeof(double));
  
	/* (X'X)^-1 */
	F77_CALL(dpotri)("U", &ngen, var, &ngen, &info);

	/* estimated variances of estimated coefficients */
	for(j=0; j<ngen; j++) wvar[j + i*ngen] = sigmasq * var[j+j*ngen];
      }

    } /* end loop over imputations */


    /* trim weights */
    for(i=0; i<ndraws; i++) index[i] = i;
    rsort_with_index(wrss, index, ndraws);
    for(i=0; i<ndraws; i++) wts[i] = exp(wts[i] - totwt);
    
    for(i=0; i<ntrim; i++) 
      wts[index[i]] = wts[index[ndraws-i-1]] = 0.0;

    /* now calculated overall estimates, 
       with var(beta) = E[ var(beta|imp) ] + var[ E(beta|imp) ] */
    for(j=0; j<ngen; j++) {
      Beta[s][j] = 0.0;
      if(getse) {
	SE[s][j] = 0.0;
	var[j] = 0.0;
      }
    }
    
    totwt = 0.0;
    for(i=0; i<ndraws; i++) {
      if(!flag[i]) {
	totwt += wts[i];

	for(j=0; j<ngen; j++) {
	  Beta[s][j] += wts[i]*wbeta[j+i*ngen];
      
	  if(getse) SE[s][j] += wts[i]*wvar[j+i*ngen];
	}
      }
    }

    for(j=0; j<ngen; j++) {
      Beta[s][j] /= totwt;
      if(getse) SE[s][j] /= totwt;
    }

    if(getse) {
      for(j=0; j<ngen; j++) {
	for(i=0; i<ndraws; i++) 
	  if(!flag[i]) 
	    var[j] += wts[i] * (wbeta[j+i*ngen] - Beta[s][j])*(wbeta[j+i*ngen] - Beta[s][j]);

	SE[s][j] += var[j] / ( (double)(ndraws-ntrim-1) * totwt / (double)(ndraws-ntrim) );

	SE[s][j] = sqrt(SE[s][j]);
      }
    }

  } /* end loop over positions */
}
 
/* end of effectscan.c */

File: R/effectscan.R
######################################################################
#
# effectscan.R
#
# copyright (c) 2003-2013, Karl W. Broman
# [completely re-written in Sep, 2007, based partly on code from Hao Wu]
# last modified Sep, 2013
# first written Jan, 2003
#
#     This program is free software; you can redistribute it and/or
#     modify it under the terms of the GNU General Public License,
#     version 3, as published by the Free Software Foundation.
#
#     This program is distributed in the hope that it will be useful,
#     but without any warranty; without even the implied warranty of
#     merchantability or fitness for a particular purpose.  See the GNU
#     General Public License, version 3, for more details.
#
#     A copy of the GNU General Public License, version 3, is available
#     at http://www.r-project.org/Licenses/GPL-3
#
# Part of the R/qtl package
# Contains: effectscan
#
######################################################################

effectscan <-
    function(cross, pheno.col=1, chr, get.se=FALSE, draw=TRUE,
             gap=25, ylim, mtick=c("line","triangle"),
             add.legend=TRUE, alternate.chrid=FALSE, ...)
{
    type <- class(cross)[1]
    mtick <- match.arg(mtick)
    if(type == "4way")
        stop("effect scan not working for 4-way cross yet.")

    if(length(pheno.col) > 1) {
        pheno.col <- pheno.col[1]
        warning("effectscan can take just one phenotype; only the first will be used")
    }

    if(is.character(pheno.col)) {
        num <- find.pheno(cross, pheno.col)
        if(is.na(num))
            stop("Couldn't identify phenotype \"", pheno.col, "\"")
        pheno.col <- num
    }

    if(pheno.col < 1 | pheno.col > nphe(cross))
        stop("pheno.col values should be between 1 and the no. phenotypes")

    if(!is.numeric(cross$pheno[,pheno.col]))
        stop("phenotype \"", colnames(cross$pheno)[pheno.col], "\" is not numeric.")

    pheno <- cross$pheno[,pheno.col]
    wh <- is.na(pheno)
    if(any(wh)) {
        pheno <- pheno[!wh]
        cross <- subset.cross(cross, ind=(!wh))
    }

    if(!missing(chr)) cross <- subset.cross(cross, chr=chr)

    chrtype <- sapply(cross$geno, class)

    n.ind <- length(pheno)

    results <- NULL
    for(i in 1:nchr(cross)) {
        if(!("draws" %in% names(cross$geno[[i]])))
            stop("You must first run sim.geno.")

        draws <- cross$geno[[i]]$draws

        # create map of positions
        if("map" %in% names(attributes(cross$geno[[i]]$draws)))
            map <- attr(cross$geno[[i]]$draws,"map")
        else {
            stp <- attr(cross$geno[[i]]$draws, "step")
            oe <- attr(cross$geno[[i]]$draws, "off.end")

            if("stepwidth" %in% names(attributes(cross$geno[[i]]$draws)))
                stpw <- attr(cross$geno[[i]]$draws, "stepwidth")
            else stpw <- "fixed"
            map <- create.map(cross$geno[[i]]$map,stp,oe,stpw)
        }

        if(is.matrix(map)) {
            marnam <- colnames(map)
            map <- map[1,]
        }
        else marnam <- names(map)

        if(type == "risib" || type=="riself" || type=="dh" || type=="haploid") {
            mapping <- rbind(c(+1, -1),
                             c(+1, +1))
            colnames(mapping) <- c("intercept","a")
            dropcol <- 1
        }
        else if(type=="bc") {
            if(chrtype[i] == "X") {
                sexpgm <- getsex(cross)
                draws <- reviseXdata(type, "full", sexpgm, draws=draws,
                                     cross.attr=attributes(cross))

                if(is.null(sexpgm$sex) || all(sexpgm$sex==0) || all(sexpgm$sex==1)) { # all one sex
                    mapping <- rbind(c(+1, -0.5),
                                     c(+1, +0.5))
                    colnames(mapping) <- c("intercept", "a")
                    dropcol <- 1
                }
                else { # some of each sex
                    mapping <- rbind(c(+1, 0,-0.5,   0),
                                     c(+1, 0,+0.5,   0),
                                     c(+1,+1, 0,  -0.5),
                                     c(+1,+1, 0,  +0.5))
                    colnames(mapping) <- c("intercept", "sex", "a.female", "a.male")
                    dropcol <- 1:2
                }
            } # end bc X chr
            else {
                mapping <- rbind(c(+1, -0.5),
                                 c(+1, +0.5))
                colnames(mapping) <- c("intercept", "a")
                dropcol <- 1
            } # end bc autosome
        } # end bc
        else { # intercross
            if(chrtype[i] == "X") {
                sexpgm <- getsex(cross)
                draws <- reviseXdata(type, "full", sexpgm, draws=draws,
                                     cross.attr=attributes(cross))


                if(is.null(sexpgm$pgm) || all(sexpgm$pgm==0) || all(sexpgm$pgm==1)) { # all one direction

                    if(is.null(sexpgm$sex) || all(sexpgm$sex==0) || all(sexpgm$sex==1)) { # all one sex
                        mapping <- rbind(c(+1, -0.5),
                                         c(+1, +0.5))
                        colnames(mapping) <- c("intercept", "a")
                        dropcol <- 1
                    }
                    else {
                        mapping <- rbind(c(+1, 0,-0.5,   0),
                                         c(+1, 0,+0.5,   0),
                                         c(+1,+1, 0,  -0.5),
                                         c(+1,+1, 0,  +0.5))
                        colnames(mapping) <- c("intercept", "sex", "a.female", "a.male")
                        dropcol <- 1:2
                    }
                }
                else { # some of each direction
                    if(is.null(sexpgm$sex) || all(sexpgm$sex==0)) { # all female
                        mapping <- rbind(c(+1, 0,-0.5,  0),
                                         c(+1, 0,+0.5,  0),
                                         c(+1,+1,   0,-0.5),
                                         c(+1,+1,   0,+0.5))
                        colnames(mapping) <- c("intercept","dir","a.forw","a.rev")
                        dropcol <- 1:2
                    }
                    else if(all(sexpgm$sex==1)) { # all male
                        mapping <- rbind(c(+1, -0.5),
                                         c(+1, +0.5))
                        colnames(mapping) <- c("intercept", "a")
                        dropcol <- 1
                    }
                    else { # some of each sex
                        mapping <- rbind(c(+1, 0, 0, -0.5,  0,   0),
                                         c(+1, 0, 0, +0.5,  0,   0),
                                         c(+1,+1, 0,   0,-0.5,   0),
                                         c(+1,+1, 0,   0,+0.5,   0),
                                         c(+1, 0,+1,   0,   0,-0.5),
                                         c(+1, 0,+1,   0,   0,+0.5))
                        colnames(mapping) <- c("intercept","dir","sex","a.femaleforw","a.femalerev","a.male")
                        dropcol <- 1:3
                    }
                }
            } # end f2 X chr
            else {
                mapping <- rbind(c(+1, -1,  0),
                                 c(+1,  0, +1),
                                 c(+1, +1,  0))
                colnames(mapping) <- c("intercept","a","d")
                dropcol <- 1
            } # f2 autosome
        } # end f2

        n.gen <- ncol(mapping)
        n.pos <- ncol(draws)
        n.imp <- dim(draws)[3]

        z <- .C("R_effectscan",
                as.integer(n.ind),
                as.integer(n.gen),
                as.integer(n.imp),
                as.integer(n.pos),
                as.integer(draws-1),
                as.double(pheno),
                as.double(mapping),
                beta=as.double(rep(0,n.pos*n.gen)),
                se=as.double(rep(0,n.pos*n.gen)),
                as.integer(get.se),
                PACKAGE="qtl")


        beta <- t(matrix(z$beta, ncol=n.pos))
        colnames(beta) <- colnames(mapping)

        if(get.se) {
            se <- t(matrix(z$se, ncol=n.pos))
            colnames(se) <- paste("se", colnames(mapping), sep=".")
            beta <- cbind(beta, se[,-dropcol,drop=FALSE])
        }
        z <- beta[,-dropcol,drop=FALSE]

        w <- marnam
        o <- grep("^loc-*[0-9]+",w)
        if(length(o) > 0) # inter-marker locations cited as "c*.loc*"
            w[o] <- paste("c",names(cross$geno)[i],".",w[o],sep="")
        rownames(z) <- w

        z <- as.data.frame(z, stringsAsFactors=TRUE)
        z <- cbind(chr=rep(names(cross$geno)[i],length(map)),
                   pos=as.numeric(map), z)
        rownames(z) <- w

        if(i==1) results <- z
        else {
            w <- match(colnames(z), colnames(results))

            if(any(is.na(w))) {
                curnam <- colnames(results)
                for(j in which(is.na(w)))
                    results <- cbind(results, rep(NA, nrow(results)))
                colnames(results) <- c(curnam, colnames(z)[is.na(w)])
            }

            w <- match(colnames(results), colnames(z))
            if(any(is.na(w))) {
                curnam <- colnames(z)
                for(j in which(is.na(w)))
                    z <- cbind(z, rep(NA, nrow(z)))
                colnames(z) <- c(curnam, colnames(results)[is.na(w)])
            }

            results <- rbind(results, z)

        }
    } # end  loop over chromosomes

    class(results) <- c("effectscan", "scanone", "data.frame")

    if(draw) { # make the figure
        if(missing(ylim))
            plot.effectscan(results, gap=gap, mtick=mtick, add.legend=add.legend,
                            alternate.chrid=alternate.chrid, ...)
        else
            plot.effectscan(results, gap=gap, mtick=mtick, add.legend=add.legend,
                            ylim=ylim, alternate.chrid=alternate.chrid, ...)
    }

    invisible(results)
}

# function to make the effectscan plot
plot.effectscan <-
    function(x, gap=25, ylim, mtick=c("line","triangle"),
             add.legend=TRUE, alternate.chrid=FALSE, ...)
{
    col <- c("blue","red","darkorange","darkgreen","purple")
    lightcol <- c("lightblue", "pink", "peachpuff1", "palegreen1", "thistle1")

    results <- x
    eff <- 3:ncol(results)

    if(length(grep("^se", colnames(results)))>0) get.se <- TRUE
    else get.se <- FALSE

    if(get.se) {
        se <- grep("^se", colnames(results)[eff])
        eff <- eff[-se]
        se <- se + 2

        lo <- as.matrix(results[,eff]) - as.matrix(results[,se])
        hi <- as.matrix(results[,eff]) + as.matrix(results[,se])

        yl <- range(c(lo,hi), na.rm=TRUE)
    }
    else yl <- range(results[,eff], na.rm=TRUE)

    if(!missing(ylim)) yl <- ylim

    plot.scanone(results, lodcolumn=1, ylim=yl, gap=gap, mtick=mtick, alternate.chrid=alternate.chrid,
                 col=col[1], ...)

    if(get.se) {
        begend <- matrix(unlist(tapply(results[,2],results[,1],range)),ncol=2,byrow=TRUE)
        rownames(begend) <- unique(results[,1])
        chr <- unique(as.character(results[,1]))

        begend <- begend[as.character(chr),,drop=FALSE]
        len <- begend[,2]-begend[,1]

        if(length(len)>1) start <- c(0,cumsum(len+gap))-c(begend[,1],0)
        else start <- 0

        x <- results[,2]
        for(i in seq(along=chr))
            x[results[,1]==chr[i]] <- results[results[,1]==chr[i],2]+start[i]

        for(i in seq(along=chr)) {
            wh <- results[,1]==chr[i]

            for(j in 1:ncol(lo)) {
                if(any(!is.na(lo[wh,j]))) {
                    xx <- c(x[wh], rev(x[wh]))
                    yy <- c(lo[wh,j], rev(hi[wh,j]))
                    polygon(xx, yy, col=lightcol[j], border=lightcol[j])
                }
            }
        }

        # go back and add lines at edges
        for(i in seq(along=chr)) {
            wh <- results[,1]==chr[i]

            for(j in 1:ncol(lo)) {
                if(any(!is.na(lo[wh,j]))) {
                    xx <- c(x[wh], rev(x[wh]))
                    yy <- c(lo[wh,j], rev(hi[wh,j]))
                    lines(xx, yy, col=lightcol[j])
                }
            }
        }

        plot.scanone(results, lodcolumn=1, add=TRUE, col=col[1])
    }

    if(length(eff) > 1) {
        for(i in seq(along=eff)[-1])
            plot.scanone(results, lodcolumn=eff[i]-2, gap=gap, add=TRUE, col=col[i])
    }

    if(add.legend)
        legend("top", legend=names(results)[eff], col=col[1:length(eff)], lwd=2)

    abline(h=0, lty=2)
}



# end of effectscan.R
